pyresample/gradient/__init__.py

```diff
@@ -178,26 +310,45 @@ def reshape_to_stacked_3d(array):
     return np.stack(layers, axis=-1)
 
 
-def parallel_gradient_search(data, src_x, src_y, dst_x, dst_y, **kwargs):
+def get_border_lonlats(geo_def, x_stride=1, y_stride=1):
+    """Get the border x- and y-coordinates."""
+    if geo_def.proj_dict['proj'] == 'geos':
+        lon_b, lat_b = get_geostationary_bounding_box(geo_def, 3600)
+    else:
+        lons, lats = geo_def.get_boundary_lonlats()
+        lon_b = np.concatenate((lons.side1, lons.side2, lons.side3, lons.side4))
+        lat_b = np.concatenate((lats.side1, lats.side2, lats.side3, lats.side4))
+
+    return lon_b, lat_b
+
+
+def get_polygon(x_borders, y_borders):
+    """Get border polygon from *x_borders* and *y_borders*."""
+    boundary = [(x_borders[i], y_borders[i]) for i in range(len(x_borders))
+                if np.isfinite(x_borders[i]) and np.isfinite(y_borders[i])]
+
+    return Polygon(boundary)
+
+
+def parallel_gradient_search(data, src_x, src_y, dst_x, dst_y,
+                             src_gradient_xl, src_gradient_xp,
+                             src_gradient_yl, src_gradient_yp, **kwargs):
     """Run gradient search in parallel in input area coordinates."""
-    if data.ndim not in [2, 3]:
-        raise NotImplementedError('Gradient search resampling only supports 2D or 3D arrays.')
-    if data.ndim == 2:
-        data = data[np.newaxis, :, :]
-    # TODO: Make sure the data is uniformly chunked.
-    src_gradient_xl, src_gradient_xp = np.gradient(src_x, axis=[0, 1])
-    src_gradient_yl, src_gradient_yp = np.gradient(src_y, axis=[0, 1])
-    arrays = reshape_arrays_in_stacked_chunks((src_x, src_y,
-                                               src_gradient_xl, src_gradient_xp, src_gradient_yl, src_gradient_yp),
-                                              src_x.chunks)
-    # TODO: rechunk and reformat the data array
-    src_x, src_y, src_gradient_xl, src_gradient_xp, src_gradient_yl, src_gradient_yp = arrays
-    data = reshape_to_stacked_3d(data)
-    res = da.blockwise(_gradient_resample_data, 'bmnz', data.astype(np.float64), 'bijz',
-                       src_x, 'ijz', src_y, 'ijz',
-                       src_gradient_xl, 'ijz', src_gradient_xp, 'ijz',
-                       src_gradient_yl, 'ijz', src_gradient_yp, 'ijz',
-                       dst_x, 'mn', dst_y, 'mn',
-                       dtype=np.float64,
-                       method=kwargs.get('method', 'bilinear'))
-    return da.nanmax(res, axis=-1).squeeze()
+    if data.ndim != 4:
+        raise NotImplementedError(
+            'Gradient search resampling only supports 4D arrays.')
+
+    if data.shape[-1] == 0:
+        logger.warning("Data doesn't cover the target area.")
+        return None
+    else:
```

Comment:
The smallest of nit-picks, this `else` doesn't have to be here since the above `if` returns `None`.